otx/algorithms/detection/tasks/task.py

```diff
@@ -0,0 +1,608 @@
+        if train_parameters:
+            update_progress_callback = train_parameters.update_progress
+        else:
+            update_progress_callback = default_progress_callback
+        self._time_monitor = TrainingProgressCallback(update_progress_callback)
+
+        results = self._train_model(dataset)
+
+        # Check for stop signal when training has stopped. If should_stop is true, training was cancelled and no new
+        if self._should_stop:
+            logger.info("Training cancelled.")
+            self._should_stop = False
+            self._is_training = False
+            return
+
+        # get output model
+        model_ckpt = results.get("final_ckpt")
+        if model_ckpt is None:
+            logger.error("cannot find final checkpoint from the results.")
+            # output_model.model_status = ModelStatus.FAILED
+            return
+        # update checkpoint to the newly trained model
+        self._model_ckpt = model_ckpt
+
+        # get prediction on validation set
+        self._is_training = False
+        val_dataset = dataset.get_subset(Subset.VALIDATION)
+        val_preds, val_map = self._infer_model(val_dataset, InferenceParameters(is_evaluation=True))
+
+        preds_val_dataset = val_dataset.with_empty_annotations()
+        self._add_predictions_to_dataset(val_preds, preds_val_dataset, 0.0)
+
+        result_set = ResultSetEntity(
+            model=output_model,
+            ground_truth_dataset=val_dataset,
+            prediction_dataset=preds_val_dataset,
+        )
+
+        # adjust confidence threshold
+        if self._hyperparams.postprocessing.result_based_confidence_threshold:
+            best_confidence_threshold = None
+            logger.info("Adjusting the confidence threshold")
+            metric = MetricsHelper.compute_f_measure(result_set, vary_confidence_threshold=True)
+            if metric.best_confidence_threshold:
+                best_confidence_threshold = metric.best_confidence_threshold.value
+            if best_confidence_threshold is None:
+                raise ValueError("Cannot compute metrics: Invalid confidence threshold!")
+            logger.info(f"Setting confidence threshold to {best_confidence_threshold} based on results")
+            self.confidence_threshold = best_confidence_threshold
+        else:
+            metric = MetricsHelper.compute_f_measure(result_set, vary_confidence_threshold=False)
+
+        # compose performance statistics
+        # TODO[EUGENE]: HOW TO ADD A MAE CURVE FOR TaskType.COUNTING?
+        performance = metric.get_performance()
+        performance.dashboard_metrics.extend(self._generate_training_metrics(self._learning_curves, val_map))
+        logger.info(f"Final model performance: {str(performance)}")
+        # save resulting model
+        self.save_model(output_model)
+        output_model.performance = performance
+        logger.info("train done.")
+
+    def _train_model(self, dataset: DatasetEntity):
```

Comment:
Let's make `@abstractmethod` decorator here if all child task.py actually do the main training procedure by this func.

Reply:
Great idea. Thanks